File: Top-Interview-Questions/Strings/Easy-Collection/38-count-and-say.js
/**
 * @param {number} n
 * @return {string}
 * Runtime: 64 ms, faster than 40.44% of JavaScript online submissions for Count and Say.
 * Memory Usage: 37.4 MB, less than 8.33% of JavaScript online submissions for Count and Say.
 * TC: O(n), but since 1<=n<=30, it is also legit to say O(1)
 * SC: O(n), but since 1<=n<=30, it is also legit to say O(1)
 */
var countAndSay = function(n) {
    let str = '1';
    while (n > 1) {
        let newStr = '', count = 0, say = str[0];
        
        for (let i = 0; i < str.length; i += 1) {
            if (str[i] === say) {
                count += 1;
            } else {
                newStr += count + say;
                count = 1;
                say = str[i];
            }
        }
        str = newStr + count + say;
        n -= 1;
    }
    return str;
};

File: Top-Interview-Questions/Strings/Easy-Collection/38-count-and-say.py
class Solution:
    def countAndSay(self, n: int) -> str:
        if n==1:
            return "1"
        
        prev = self.countAndSay(n-1)
        ret = ""
        count = 1
        for i in range(len(prev)):
            if i == len(prev)-1 or prev[i] != prev[i+1]: # we cannot compare the last element
                ret = ret+str(count) + prev[i] # same as ret += str(count) + prev[i]
                count = 1 # reset to 1 to find other numbers' counts
            else:
                count += 1
        return ret
        # let's find the rule of each number
        # 1: 1
        # 2: 11, one of element 1
        # 3: 21, two of element 1
        # 4: 1211, one of element 2, one of element one
        # So, it basically sees the previous number to find the pattern
        # 5: 111221, one of element 1, one of element 2, two of element 1
        # 6: 312211, three of element 1, two of element 2, one of element 1
        # 7: 13112221, ...

        # TC: O(n), O(1) if we say 1<=n<=30
        # SC: O(n), string slicing, O(1) if we say 1<=n<=30
        # Runtime: 20 ms, faster than 99.95%
        # Memory: 13.9 MB, less than 6.38% 